Bottom: bcd5cdb833e385404b5d9f7a62d720a03c1231b6
Top:    4e7c3030c1590ad5fe2cfaf9df12bc72e70b6452
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-07-25 20:20:06 +0800

Refresh of tun-xdp-support

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 3c9e60e..deb446d 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1270,17 +1270,11 @@ static void tun_rx_batched(struct tun_struct *tun, struct tun_file *tfile,
 static void tun_xdp_xmit(struct tun_struct *tun, struct tun_file *tfile,
 			 struct sk_buff *skb)
 {
-	int ret;
-
 	skb->queue_mapping = tfile->queue_index;
+	skb->protocol = eth_type_trans(skb, tun->dev);
 	local_bh_disable();
-	ret = tun_net_xmit(skb, tun->dev);
+	tun_net_xmit(skb, tun->dev);
 	local_bh_enable();
-	if (ret == NET_XMIT_DROP) {
-		this_cpu_inc(tun->pcpu_stats->tx_dropped);
-		kfree_skb(skb);
-	}
-
 	return;
 }
 
@@ -1361,6 +1355,7 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	if (generic_xdp) {
 		int ret;
 
+		/* FIXME: hdr conversion and other stuffs */
 		rcu_read_lock();
 		ret = do_xdp_generic(rcu_dereference(tfile->xdp_prog), skb);
 		rcu_read_unlock();
